### PR TITLE
Add Future.bichain

### LINKED ISSUE #332: Add a bichain function
```hs
bichain :: (a -> Future x y) -> (b -> Future x y) -> Future a b -> Future x y
```

----

Currently the only way to bichain is like this:

```js
compose (chain (x => x)) (coalesce (f) (g))
```

It's not straightforward and one might be inclined to do something like this instead:

```js
compose (chainRej (f)) (chain (g))
```

However, this solution behaves different from how `bichain` would, in that if `g` returns a *rejected* Future, its rejection is handled by `f`.

### PR DESCRIPTION
**Pull request by dicearr**
Close #332

### CODE REVIEW

**Comment by dicearr** [comment]
I don't know to which _algebraic properties for the left branch of bichain_ you're referring to.

I was also thinking of adding a `chain derived from bichain` test in the same way `map derived from bimap` although, bichain is not part of the Fantasy Land spec.

File: test/prop/0.algebra.js [hunk 1]
```diff
@@ -54,6 +55,18 @@ property('ap composition using map', _mx, _mf, _mf, function (mx, mf, mg){
   return eq(ap(mx)(ap(mf)(map(B)(mg))))(ap(ap(mx)(mf))(mg));
 });
 
+property('bichain associativity', _mx, _fm, _fm, function (m, f, g){
+  return eq(bichain(g)(g)(bichain(f)(f)(m)))(bichain(B(bichain(g)(g))(f))(B(bichain(g)(g))(f))(m));
+});
+
+property('bichain left identity', _x, _fm, function (x, f){
+  return eq(bichain(f)(reject)(reject(x)))(f(x));
+});
```
[the remark this answers is not in the document]
Comment (Avaq) [reply test/prop/0.algebra.js]: Essentially what we're testing is that `chain (resolve)` is the identity function for `m`. So what you should be testing is that `bichain (reject) (resolve)` is the identity function for `m`.

[the remark this answers is not in the document]
Comment (Avaq) [reply test/prop/0.algebra.js]: In the `chain` equations, this property is verified in two steps:

```js
left  | chain (f) (resolve (x)) = f (x)
right | chain (resolve) (m)     = m
```

[the remark this answers is not in the document]
Comment (Avaq) [reply test/prop/0.algebra.js]: So for bichain it should probably look like:

```js
left  | bichain (f) (g) (reject (x))   = f (x)
left  | bichain (f) (g) (resolve (x))  = g (x)
right | bichain (reject) (resolve) (m) = m
```

[the remark this answers is not in the document]
Comment (dicearr) [reply test/prop/0.algebra.js]: I don't understand now the difference between _right_ and _left_ identity. I always thought it was related to the branches. In any case, from your comments I have changed the code and hopefully it is right now.

[the remark this answers is not in the document]
Comment (dicearr) [reply test/prop/0.algebra.js]: I don't understand now the difference between _right_ and _left_ identity. I always thought it was related to the branches. In any case, from your comments I have changed the code and hopefully it is right now.

[the remark this answers is not in the document]
Comment (Avaq) [reply test/prop/0.algebra.js]: I don't really understand the [naming of the assertions](https://github.com/fantasyland/fantasy-land/commit/f5a70482706ab8eb628eba61b2409bde6ca2fba3) either. Maybe the idea was something like: "identity as asserted from the left side" and "identity as asserted from the right side". Maybe we can get @puffnfresh to recall how he got to the naming, and whether our interpretation of it is right.

I think of the two assertions as:

1. "given that `chain (f) (resolve (x))` is the equivalent of `f (x)`"; (left identity)
2. "`chain (resolve) (m)` should be the equivalent of `m`". (right identity)

But I might be wrong. The code looks good to me now, but let's wait a bit and see if Brian will respond :)

**Pull request merged**